lib/chef/provider/aws_cloudsearch_domain.rb

```diff
@@ -0,0 +1,126 @@
+require 'chef/provisioning/aws_driver/aws_provider'
+
+class Chef::Provider::AwsCloudsearchDomain < Chef::Provisioning::AWSDriver::AWSProvider
+
+  def create_aws_object
+    converge_by "Creating new CloudSearch domain #{new_resource.domain_name}" do
+      create_domain
+    end
+    # multi_az defaults to false on creation
+    # so we only need to hit the API if it is true
+    if new_resource.multi_az
+      converge_by "Setting availability options for CloudSearch domain #{new_resource}" do
+        update_availability_options
+      end
+    end
+
+    if update_scaling_params?
+      converge_by "Setting scaling parameters for CloudSearch domain #{new_resource.domain_name}" do
+        update_scaling_parameters
+      end
+    end
+
+    if update_policy?
+      converge_by "Setting access policy for CloudSearch domain #{new_resource.domain_name}" do
+        update_service_access_policy
+      end
+    end
+
+    if update_index_fields?
+      converge_by "Defining index fields for CloudSearch domain #{new_resource.domain_name}" do
+        new_resource.index_fields.each do |idx_field|
+          if !idx_field.is_a? Hash
+            raise ArgumentError, "Expected index_fields to contain only Hashes"
+          else
+            create_index_field(idx_field)
+          end
+        end
+      end
+    end
+  end
+
+  def destroy_aws_object(instance)
+    converge_by "Deleting CloudSearch domain #{new_resource.domain_name}" do
+      cs_client.delete_domain(domain_name: new_resource.domain_name)
+    end
+  end
+
+  def update_aws_object(instance)
+    if update_availability_options?
+      converge_by "Updating availability options for CloudSearch domain #{new_resource}" do
+        update_availability_options
+      end
+    end
+
+    if update_scaling_params?
+      converge_by "Updating scaling parameters for CloudSearch domain #{new_resource.domain_name}" do
+        update_scaling_parameters
+      end
+    end
+
+    if update_policy?
+      converge_by "Updating access policy for CloudSearch domain #{new_resource.domain_name}" do
+        update_service_access_policy
+      end
+    end
+
+    if update_index_fields?
+      Chef::Log.warn("Updating existing index_fields not currently supported")
+    end
+  end
+
+
+  # TODO(ssd): Expand these to be idempotent?  It takes an extra API
```

Comment:
I vote to make them idempotent.  After the create is finished we will know the current state.  And the `update_aws_object` is passed the current aws object, so this query will have already been made.  These `?` methods just need to take in that existing object for comparison with `new_resource`.

Man, this will be sooooooo much easier with resources 2.0